pkg/egressgateway/manager.go

```diff
@@ -265,23 +355,63 @@ func (manager *Manager) OnUpdateEndpoint(endpoint *k8sTypes.CiliumEndpoint) {
 		return
 	}
 
+	if _, ok := manager.epDataStore[epData.id]; ok {
+		logger.Debug("Updated CiliumEndpoint")
+	} else {
+		logger.Debug("Added CiliumEndpoint")
+	}
+
 	manager.epDataStore[epData.id] = epData
 
 	manager.reconcile(eventUpdateEndpoint)
 }
 
+// OnUpdateEndpoint is the event handler for endpoint additions and updates.
+func (manager *Manager) OnUpdateEndpoint(endpoint *k8sTypes.CiliumEndpoint) {
+	id := types.NamespacedName{
+		Name:      endpoint.GetName(),
+		Namespace: endpoint.GetNamespace(),
+	}
+
+	// remove the ID from the workqueue in case an older revision of the
+	// same CiliumEndpoint was already present
+	manager.endpointEventRetryQueue.Forget(id)
+	manager.endpointEventRetryQueue.Done(id)
+
+	manager.Lock()
+	manager.pendingEndpointEvents[id] = endpoint
+	manager.Unlock()
+
+	manager.addEndpoint(id)
+
+	if manager.endpointEventIsPending(id) {
+		manager.endpointEventRetryQueue.Add(id)
+	}
+}
+
 // OnDeleteEndpoint is the event handler for endpoint deletions.
 func (manager *Manager) OnDeleteEndpoint(endpoint *k8sTypes.CiliumEndpoint) {
 	manager.Lock()
 	defer manager.Unlock()
 
+	logger := log.WithFields(logrus.Fields{
+		logfields.K8sEndpointName: endpoint.Name,
+		logfields.K8sNamespace:    endpoint.Namespace,
+	})
+
 	id := types.NamespacedName{
 		Name:      endpoint.GetName(),
 		Namespace: endpoint.GetNamespace(),
 	}
 
 	delete(manager.epDataStore, id)
 
+	delete(manager.pendingEndpointEvents, id)
+	manager.endpointEventRetryQueue.Forget(id)
+	manager.endpointEventRetryQueue.Done(id)
```

Comment:
Same here about just letting the event to be consumed by the reconcile loop rather than removing it explicitly.